sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,374 @@
+ * the lower half of the actual register.
+ * @param upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param mask Mask that identifies the bits to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                              uint32_t upper_offset, uint32_t mask,
+                              uint32_t val) {
+  const uint32_t mask_lower_half = mask & 0x0000FFFFu;
+  const uint32_t mask_upper_half = mask & 0xFFFF0000u;
+  if (mask_lower_half != 0) {
+    mmio_region_write32(gpio->base_addr, lower_offset,
+                        (mask_lower_half << 16) | (val & 0x0000FFFFu));
+  }
+  if (mask_upper_half != 0) {
+    mmio_region_write32(gpio->base_addr, upper_offset,
+                        mask_upper_half | ((val & 0xFFFF0000u) >> 16));
+  }
+}
+
+/**
+ * Perform a masked write to a single bit of a GPIO register.
+ *
+ * The GPIO device provides masked bit-level atomic writes to its DIRECT_OUT
+ * and DIRECT_OE registers. This allows software to modify half of the bits
+ * at a time without requiring a read-modify-write. This function is guaranteed
+ * to perform only one write since it never needs to access both halves of a
+ * register.
+ *
+ * See also |gpio_masked_write()|.
+ *
+ * @param gpio GPIO instance.
+ * @param lower_offset Offset of the masked access register that corresponds to
+ * the lower half of the actual register.
+ * @param upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param index Zero-based index of the bit to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_bit_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                                  uint32_t upper_offset, uint32_t index,
+                                  bool val) {
+  // Write to lower_offset if the bit is in the lower half, write to
+  // upper_offset otherwise.
+  const ptrdiff_t offset = (index < 16) ? lower_offset : upper_offset;
+  // Since masked access writes to half of a register, index mod 16 gives the
+  // index of the bit in the half-word mask.
+  const uint32_t mask = index_to_mask(index % 16);
+  mmio_region_write32(gpio->base_addr, offset,
+                      (mask << 16) | (val ? mask : 0u));
+}
+
+dif_gpio_result_t dif_gpio_init(const dif_gpio_config_t *config,
+                                dif_gpio_t *gpio) {
+  if (config == NULL || gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  // Save internal state in the given |dif_gpio_t| instance.
+  gpio->base_addr = config->base_addr;
+  // Reset the GPIO device at the given |base_addr|.
+  dif_gpio_reset(gpio);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_reset(const dif_gpio_t *gpio) {
+  if (gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
```

Comment:
Thanks for catching this! While I would expect interrupt handlers to care of the pending interrupts, it is a good idea to clear them as part of reset to avoid any surprises.